src/components/plots/TrajectoryAnalysisPlot.jsx

```diff
@@ -0,0 +1,147 @@
+
+import { getCellSets } from 'redux/selectors';
+import PlatformError from 'components/PlatformError';
+import Loader from 'components/Loader';
+import changeEmbeddingAxesIfNecessary from 'components/plots/helpers/changeEmbeddingAxesIfNecessary';
+
+const TrajectoryAnalysisPlot = (props) => {
+  const {
+    experimentId,
+    config,
+    plotData,
+    actions,
+    onUpdate,
+  } = props;
+  const dispatch = useDispatch();
+
+  const cellSets = useSelector(getCellSets());
+
+  const embeddingSettings = useSelector(
+    (state) => state.experimentSettings.originalProcessing?.configureEmbedding?.embeddingSettings,
+  );
+
+  const {
+    data: embeddingData,
+    loading: embeddingLoading,
+    error: embeddingError,
+  } = useSelector(
+    (state) => state.embeddings[embeddingSettings?.method],
+  ) || {};
+
+  const [plotSpec, setPlotSpec] = useState({});
+
+  useEffect(() => {
+    if (!embeddingSettings) {
+      dispatch(loadProcessingSettings(experimentId));
+    }
+
+    if (cellSets.loading && !cellSets.error) {
+      dispatch(loadCellSets(experimentId));
+    }
+
+    if (!embeddingData && embeddingSettings?.method) {
+      dispatch(loadEmbedding(experimentId, embeddingSettings?.method));
+    }
+  }, [experimentId, embeddingSettings?.method]);
+
+  useEffect(() => {
+    changeEmbeddingAxesIfNecessary(config, embeddingSettings?.method, onUpdate);
+  }, [config, embeddingSettings?.method]);
+
+  useEffect(() => {
+    if (
+      !config
+      || cellSets.loading
+      || cellSets.error
+      || !embeddingData?.length
+      || !plotData
+      || !plotData?.nodes
+    ) {
+      return;
+    }
+
+    const {
+      plotData: plotEmbedding,
+      cellSetLegendsData,
+    } = generateCategoricalEmbeddingData(cellSets, config.selectedSample, config.selectedCellSet, embeddingData);
+    const trajectoryData = generateTrajectoryPathData(plotData);
+
+    setPlotSpec(generateSpec(config, plotEmbedding, trajectoryData, cellSetLegendsData));
+  }, [config, cellSets, embeddingData, plotData]);
+
+  const render = () => {
+    if (cellSets.error) {
+      return (
+        <PlatformError
+          error={cellSets.error}
+          onClick={() => { dispatch(loadCellSets(experimentId)); }}
+        />
+      );
+    }
+
+    if (embeddingError) {
+      return (
+        <PlatformError
+          error={embeddingError}
+          onClick={() => { dispatch(loadEmbedding(experimentId, embeddingSettings?.method)); }}
+        />
+      );
+    }
+
+    if (!config
+      || cellSets.loading
+      || !embeddingData
+      || embeddingLoading
+      || Object.keys(plotSpec).length === 0
+    ) {
+      return (
+        <center>
+          <Loader experimentId={experimentId} />
+        </center>
+      );
+    }
+
+    return (
+      <center>
+        <Vega spec={plotSpec} renderer='canvas' actions={actions} />
+      </center>
+    );
+  };
+
+  return (
+    <>
+      {render()}
+    </>
```

Comment:
What's the difference between this and just `return render()`?

Reply:
It's the same. `return render()` should be simpler,